File: PHP/HELML/src/HELML.php
<?php
/*
 * This code represents a PHP implementation of the HELML class without dependencies.
 * 
 * The class provides functions for encoding and decoding data in the HELML format.
 * 
 * HELML (Header-Like Markup Language) is a markup language similar to the HTTP header format.
 * (HTTP headers are a specific case of the HELML format, a simple single-level array case).
 * 
 * HELML allows encoding arrays of any complexity
 *  and then decoding them while preserving data types for true, false, null, integer.
 *  The format is ideologically close to YAML, JSON, and other array serialization formats,
 *  intended for use as an alternative in cases where it is convenient.
 *
 * The main advantage of the HELML format is its simplicity, clarity, and minimalism.
 * The format works with any data and does not stumble on binary blocks or special characters.
 * At the same time, the format is intuitively understandable, can be created and edited manually as text.
 * In most cases, data in HELML will be more compact than in other markup languages.
 * 
 * This class implements two varieties of HELML: multiline format and URL format.
 * 
 * The URL variety of HELML is intended for data transmission in the URL string,
 * for example, to implement APIs in GET requests. In this case, the format packs data
 * into a single line and minimizes the number of "inconvenient" characters for the URL.
 *
 * Main methods of the class:
 * encode: takes an array of data and converts it into a string using the HELML format.
 * the second parameter of the method determines whether to use the URL mode.
 *
 * decode: performs the reverse transformation of encode.
 * takes a HELML formatted string and converts it back into an array.
 * automatically determines whether the input data is in URL format or multiline format.
 * 
 * valueEncoder: an internal method that encodes a value (not an array) in the HELML format
 * valueDecoder: an internal method that decodes HELML values back into their original data.
 * 
 * All methods are static and can be used without creating an instance of the class.
*/

class HELML {
    public static function encode($arr, $url_mode = false, $val_encoder = true) {
        $results_arr = [];
        if (!is_array($arr)) {
            throw new InvalidArgumentException("Array required");
        }
        $str_imp = $url_mode ? "~" : "\n";
        $lvl_ch = $url_mode ? '.' : ':';
        $spc_ch = $url_mode ? '_' : ' ';
        self::_encode($arr, $results_arr, $val_encoder, 0, $lvl_ch, $spc_ch);
        return implode($str_imp, $results_arr);
    }

    public static function _encode(
        $arr,
        &$results_arr,
        $val_encoder = true,
        $level = 0,
        $lvl_ch = ":",
        $spc_ch = " "
    ) {
        foreach ($arr as $key => $value) {

            // encode $key in base64url if it contains unwanted characters
            $fc = substr($key, 0, 1);
            $lc = substr($key, -1, 1);
            if (false !== strpos($key, $lvl_ch) || false !== strpos($key, '~') || '#' === $fc  || $fc === $spc_ch || $fc === ' ') {
                $fc = "-";
            }
            if ("-" === $fc || $lc === $spc_ch || $lc === ' ' || !preg_match('/^[[:print:]]*$/', $key)) {
                // add "-" to the beginning of the key to indicate it's in base64url
                $key = "-" . self::base64Uencode($key);
            }

            // add the appropriate number of colons to the left of the key, based on the current level
            $key = str_repeat($lvl_ch, $level) . $key;

            if (is_array($value)) {
                // if the value is an array, call this function recursively and increase the level
                $results_arr[] = $key;
                self::_encode($value, $results_arr, $val_encoder, $level + 1, $lvl_ch, $spc_ch);
            } else {
                // if the value is not an array, run it through a value encoding function, if one is specified
                if (true === $val_encoder) {
                    $value = self::valueEncoder($value, $spc_ch); // Default value encoder
                } elseif ($val_encoder) {
                    $value = call_user_func($val_encoder, $value);
                }
                // add the key:value pair to the output
                $results_arr[] = $key . $lvl_ch . $value;
            }
        }
    }

    public static function decode($src_rows, $val_decoder = true) {
        // If the input is an array, use it. Otherwise, split the input string into an array.
        if (is_array($src_rows)) {
            $str_arr = $src_rows;
        } elseif (is_string($src_rows)) {
            foreach(["\n", "\r", "~"] as $exploder_ch) {
                if (false !== strpos($src_rows, $exploder_ch)) {
                    if ("~" === $exploder_ch) {
                        $lvl_ch = '.';
                        $spc_ch = '_';
                    } else {
                        $lvl_ch = ':';
                        $spc_ch = ' ';
                    }
                    $str_arr = explode($exploder_ch, $src_rows);
                    break;
                }
            }
        } else {
            throw new InvalidArgumentException("Array or String required");
        }

        // Initialize result array and stack for keeping track of current array nesting
        $result = [];
        $stack = [];

        // Loop through each line in the input array
        foreach ($str_arr as $line) {
            $line = trim($line);

            // Skip empty lines and comment lines starting with '#'
            if (!strlen($line) || substr($line, 0, 1) === '#') continue;

            // Calculate the level of nesting for the current line by counting the number of colons at the beginning
            $level = 0;
            while (substr($line, $level, 1) === $lvl_ch) {
                $level++;
            }

            // If the line has colons at the beginning, remove them from the line
            if ($level) {
                $line = substr($line, $level);
            }

            // Split the line into a key and a value (or null if the line starts a new array)
            $parts = explode($lvl_ch, $line, 2);
            $key = $parts[0] ? $parts[0] : 0;
            $value = isset($parts[1]) ? $parts[1] : null;

            // Decode the key if it starts with an equals sign
            if (is_string($key) && ('-' === substr($key, 0, 1))) {
                $key = self::base64Udecode(substr($key, 1));
                if (!$key) {
                    $key = "ERR";
                }
            }

            // Remove keys from the stack until it matches the current level
            while (count($stack) > $level) {
                array_pop($stack);
            }

            // Find the parent element in the result array for the current key
            $parent = &$result;
            foreach ($stack as $parentKey) {
                $parent = &$parent[$parentKey];
            }

            // If the value is null, start a new array and add it to the parent array
            if (is_null($value)) {
                $parent[$key] = [];
                array_push($stack, $key);
            } else {
                // Decode the value if a decoder function is specified
                if (true === $val_decoder) {
                    $value = self::valueDecoder($value, $spc_ch);
                } elseif ($val_decoder) {
                    $value = call_user_func($val_decoder, $value, $spc_ch);
                }
                // Add the key-value pair to the current array
                $parent[$key] = $value;
            }
        }

        // Return the result array
        return $result;
    }
    
    public static function valueEncoder($value, $spc_ch = ' ') {
        $type = gettype($value);
        switch ($type) {
            case 'string':
                if ('_' === $spc_ch) {
                    // for url-mode
                    $need_encode = (false !== strpos($value, '~'));
                    $reg_str = '/^[ -~]*$/';
                } else {
                    $need_encode = false;
                    $reg_str = '/^[[:print:]]*$/u';
                }
                if ($need_encode || !preg_match($reg_str, $value) || (('_' === $spc_ch) && (false !== strpos($value, '~')))) {
                    // if the string contains special characters, encode it in base64
                    return self::base64Uencode($value);
                } elseif (!strlen($value) || ($spc_ch === $value[0]) || ($spc_ch == substr($value, -1)) || ctype_space(substr($value, -1))) {
                    // for empty strings or those that have spaces at the beginning or end
                    return '"' . $value . '"';
                } else {
                    // if the value is simple, just add one space at the beginning
                    return $spc_ch . $value;
                }
            case 'boolean':
                return $spc_ch . $spc_ch . ($value ? 'T' : 'F');
            case 'NULL':
                return $spc_ch . $spc_ch . 'N';
            case 'double':
                if ('_' === $spc_ch) {
                    // for url-mode because dot-inside
                    return self::base64Uencode($value);
                }
                // if not url mode, go below
            case 'integer':
                return $spc_ch . $spc_ch . $value;
            default:
                throw new InvalidArgumentException("Cannot encode value of type $type");
        }
    }

    public static function valueDecoder($encodedValue, $spc_ch = ' ') {
        $fc = substr($encodedValue, 0, 1);
        if ($spc_ch === $fc) {
            if (substr($encodedValue, 0, 2) !== $spc_ch . $spc_ch) {
                // if the string starts with only one space, return the string after it
                return substr($encodedValue, 1);
            }
            // if the string starts with two spaces, then it encodes a non-string value
            $encodedValue = substr($encodedValue, 2); // strip left spaces
            if ($encodedValue === 'N') {
                return null;
            } elseif ($encodedValue === 'T') {
                return true;
            } elseif ($encodedValue === 'F') {
                return false;
            }
            if (is_numeric($encodedValue)) {
                // it's probably a numeric value
                if (strpos($encodedValue, '.')) {
                    // if there's a decimal point, it's a floating point number
                    return (double) $encodedValue;
                } else {
                    // if there's no decimal point, it's an integer
                    return (int) $encodedValue;
                }
            }
            // other encoding options are not currently supported
            return $encodedValue;
        } elseif ('"' === $fc || "'" === $fc) { // it's likely that the string is enclosed in single or double quotes
            $encodedValue = substr($encodedValue, 1, -1); // trim the presumed quotes at the edges and return the interior
            if ("'" === $fc) {
                return $encodedValue;
            }
            return stripcslashes($encodedValue)
        }
        // if there are no spaces or quotes at the beginning, the value should be in base64
        $decoded = self::base64Udecode($encodedValue)
        if (false === $decoded) {
            return $encodedValue; // Fallback if can't decode
        }
        return $decoded;
    }

    public static function base64Uencode($str) {
        $enc = base64_encode($str);
        return rtrim(strtr($enc, '+/', '-_'), '=');
    }
    
    public static function base64Udecode($str) {
        return base64_decode(strtr($str, '-_', '+/'));
    }
}

File: Python/HELML/_helml.py
import base64
import re
from typing import List, Dict, Set, Union, Callable

class HELML:

    SPEC_TYPE_VALUES = {
        'N': None,
        'U': None,
        'T': True,
        'F': False,
        'NAN': float('nan'),
        'INF': float('inf'),
        'NIF': float('-inf')
    }

    CUSTOM_FORMAT_DECODER = None
    CUSTOM_VALUE_ENCODER = None
    CUSTOM_VALUE_DECODER = None
    
    ENABLE_SPC_IDENT = 1 # For encode: add space-indentation at begin of string
    ENABLE_BONES = True # For encode: enable use "next"-keys like :--:
    ENABLE_KEY_UPLINES = True # For encode: add empty line before create-array-keys
    ENABLE_HASHSYMBOLS = True # For encode: adding # after nested-blocks
    ADD_PREFIX = False
    ADD_POSTFIX = False

    @staticmethod
    def encode(
        arr: Union[dict, list, tuple, set],
        one_line_mode: int = 0
    ) -> str:
        """
        Encode array to HELML string.

        :param arr: Input data structure (list, dict, or tuple) to be encoded.
        :param one_line_mode: The encoding mode to use (0-multi-line, 1-URL-mode, 2-LINE-mode)
        :return: Encoded HELML string.
        """
        results_arr = []

        str_imp = "~" if one_line_mode else "\n"
        url_mode = one_line_mode == 1
        lvl_ch = "." if url_mode else ":"
        spc_ch = "=" if url_mode else " "

        is_list = HELML.num_keys_cnt(arr)
        
        HELML._encode(arr, results_arr, 0, lvl_ch, spc_ch, is_list)

        if lvl_ch != ':' or spc_ch != ' ' or HELML.ADD_POSTFIX:
            results_arr.append('#' + lvl_ch + spc_ch + '~')
        elif one_line_mode:
            # Remove whitespace from beginning and end of each string
            results_arr = [s.strip() for s in results_arr]
            # Remove empty strings and comments that start with '#'
            results_arr = [s for s in results_arr if s and not s.startswith('#')]

        return str_imp.join(results_arr)
    
    @staticmethod
    def _encode(
        arr: Union[dict, list, tuple, set],
        results_arr: list,
        level: int = 0,
        lvl_ch: str = ":",
        spc_ch: str = " ",
        is_list: int = 0
    ) -> None:

        # select value-encoder function: custom or internal default
        value_enco_fun = HELML.CUSTOM_VALUE_ENCODER if HELML.CUSTOM_VALUE_ENCODER is not None else HELML.valueEncoder

        # convert arr to dict if need
        if not isinstance(arr, dict):
            arr = {index: value for index, value in enumerate(arr)}

        for key, value in arr.items():
            if not isinstance(key, str):
                key = str(key)

            if is_list and HELML.ENABLE_BONES and spc_ch == " ":
                key = '--'
            elif not is_list:
                # encode key to base64url if contains unwanted characters
                # check first char
                fc = key[:1]
                # encode key in base64url if it contains unwanted characters
                if lvl_ch in key or fc == "#" or fc == spc_ch or fc == ' ' or fc == ''  or key[-1] == spc_ch or key[-1] == ' ':
                    fc = '-'
                elif level == 0 and fc == '/' and key.startswith('//'): # prevent '//' at begin of the line
                    fc = '-'
                elif not (re.match(r'^[ -}]+$', key) if spc_ch == '_' else re.match(r'^[^\x00-\x1F\x7E-\xFF]+$', key)):
                    fc = '-'
                if fc == '-':
                    # add "-" to the beginning of the key to indicate it's in base64url
                    key = '-' + HELML.base64Uencode(key)

            # add the appropriate number of colons to the left of the key, based on the current level
            ident = lvl_ch * level

             # add space-ident to the left of the key (if need)
            if HELML.ENABLE_SPC_IDENT and spc_ch == ' ':
                ident = spc_ch * (level * HELML.ENABLE_SPC_IDENT) + ident

            is_numkeys = isinstance(value, (list, tuple, set))

            if is_numkeys or isinstance(value, dict):
                # Add empty line before create-key
                if HELML.ENABLE_KEY_UPLINES and spc_ch == ' ':
                    results_arr.append('')

                # add ":" after key for lists
                if not is_numkeys:
                    key += lvl_ch

                results_arr.append(ident + key)

                HELML._encode(value, results_arr, level + 1, lvl_ch, spc_ch, is_numkeys)

                if HELML.ENABLE_KEY_UPLINES and spc_ch == ' ':
                    results_arr.append(spc_ch * level + '#')
            else:
                # use selected value encoder function
                value = value_enco_fun(value, spc_ch)

                # add the key:value pair to the output
                results_arr.append(ident + key + lvl_ch + value)

    @staticmethod
    def decode(
        src_rows: str,
        get_layers: Union[int, str, Set[Union[str, int]], List[Union[str, int]]] = [0]
    ) -> Dict:

        # Modify get_layers if needed: convert single T to array [0, T]
        if isinstance(get_layers, (int, str)):
            get_layers = ['0', get_layers]
        # Prepare layers_set from get_layers (str)
        layers_set = set()
        for i in get_layers:
            if not isinstance(i, str):
                i = str(i)
            layers_set.add(i)

        lvl_ch: str = ":"
        spc_ch: str = " "

        # Search postfix
        postfixIndex = src_rows.find('~#'); #postfix "~#: ~"
        if postfixIndex >= 0 and src_rows[postfixIndex+4] == '~':
            # get control-chars from postfix
            lvl_ch = src_rows[postfixIndex+2]
            spc_ch = src_rows[postfixIndex+3]

            # skip prefix
            stpos = 0
            for ch in src_rows:
                if ch != ' ' and ch != "\t" and ch != '~':
                    break
                stpos += 1

            # cut string between prefix and postfix
            src_rows = src_rows[stpos:postfixIndex]

        # Explode src_rows to lines
        for exploder_ch in ["\r\n", "\r", "\n"]:
            if exploder_ch in src_rows:
                break
        
        # Replace all ~ to line divider
        src_rows = src_rows.replace("~", "\n")

        # Explode string to lines
        str_arr = src_rows.split(exploder_ch)

        return HELML._decode(str_arr, layers_set, lvl_ch, spc_ch)

    def _decode(
        str_arr: list,
        layers_set: set,
        lvl_ch: str,
        spc_ch: str
    ) -> Dict:
        # select value decoder function custom or internal default
        value_deco_fun = HELML.CUSTOM_VALUE_DECODER if HELML.CUSTOM_VALUE_DECODER is not None else HELML.valueDecoder
        
        # layer values prepare
        layer_init: str = '0'
        layer_curr:str = layer_init
        all_layers = set(['0'])

        # Initialize result array and stack for keeping track of current array nesting
        result = {}
        stack = []

        # array of stack stamps for delayed conversion dict to list
        tolist = []

        min_level: int = -1

        # Loop through each line in the input array
        for line in str_arr:
            line = line.strip()

            # Skip empty lines and comment lines starting with '#'
            if not len(line) or line[0] == "#" or line.startswith('//'):
                continue

            # Calculate the level of nesting for the current line by counting the number of colons at the beginning
            level = 0
            while line[level] == lvl_ch:
                level += 1

            # If the line has colons at the beginning, remove them from the line
            if level:
                line = line[level:]

            # Split the line into a key and a value (or null if the line starts a new array)
            parts = line.split(lvl_ch, 1)
        
            key = parts[0] if parts[0].strip() else ""
            value = parts[1] if len(parts) > 1 else None

            if len(key) == 0:
                continue

            # check min_level
            if min_level < 0 or min_level > level:
                min_level = level

            extra_keys_cnt: int = len(stack) - (level - min_level)
            if extra_keys_cnt > 0:
                layer_curr = layer_init
                while len(stack) and extra_keys_cnt > 0:
                    stack.pop()
                    extra_keys_cnt -= 1

            # Find the parent element in the result dictionary for the current key
            parent = result
            for parent_key in stack:
                parent = parent[parent_key]

            # Decode the key if it starts with an equals sign
            if key.startswith("-"):
                # Next number keys
                if key == '--':
                    # Next Num keys
                    key = (str)(len(parent))

                # Layer control keys
                elif key == '-+' or key == '-++' or key == '---':
                    if not value is None:
                        value = value.strip()
                    if key == '-++':
                        layer_init = value if value else '0'
                        layer_curr = layer_init
                    elif key == '-+':
                        if value is None:
                            layer_curr = str(int(layer_curr) + 1) if layer_curr.isdigit() else layer_init
                        else:
                            layer_curr = layer_init if value == '' else value

                    all_layers.add(layer_curr)
                    continue
                else:
                    decoded_key = HELML.base64Udecode(key[1:])
                    if decoded_key != None:
                        key = decoded_key

            # If the value is null, start a new dictionary and add it to the parent dictionary
            if value is None or value == '':
                layer_curr = layer_init
                parent[key] = {}
                stack.append(key)
                if value is None:
                    tolist.append(stack.copy())
            elif layer_curr in layers_set:
                # Decode the value by selected value-decoder-function
                # Add the key-value pair to the current dictionary
                parent[key] = value_deco_fun(value, spc_ch)

        # try convert nested arrays by keys-pathes from tolist
        for stack in tolist:
            parent = result
            for parent_key in stack[:-1]:
                parent = parent[parent_key]

            last_key = stack[-1]
            if isinstance(parent, list) and HELML.is_numeric(last_key) and int(last_key) < len(parent):
                last_key = int(last_key)
            if ((isinstance(parent, dict) and parent.get(last_key) is not None) or isinstance(last_key, int)) and isinstance(parent[last_key], dict):
                # convert from dict to list if possible
                converted = []
                keys = parent[last_key].keys()
                for i in range(len(keys)):
                    if str(i) not in keys or parent[last_key][str(i)] is None:
                        # Key not found
                        break
                    converted.append(parent[last_key][str(i)])
                else:
                    # overwrite the key only if the conversion was successful
                    parent[last_key] = converted
                

        if (len(all_layers) > 1):
            result['_layers'] = all_layers

        # Return the result dictionary
        return result

    @staticmethod
    def valueEncoder(
        value: Union[str, int, float, bool, None],
        spc_ch: str = " "
    ) -> str:
        """
        Encodes a value into a HELML string.

        Args:
            value: The value to be encoded.
            spc_ch: The space character used for encoding.

        Returns:
            str: The encoded HELML string.
        """

        value_type = type(value).__name__
        if value_type == "str":
            if not value or not (re.match(r'^[ -}]+$', value) if spc_ch == '_' else re.match(r'^[^\x00-\x1F\x7E-\xFF]+$', value)):
                # if the string contains special characters, encode it in base64
                return '-' + HELML.base64Uencode(value)
            elif value[0] == spc_ch or value[-1] == spc_ch or value[-1] == ' ':
                # for empty strings or those that have spaces at the beginning or end
                return "'" + value + "'"

            # if the value is simple, just add one space at the beginning
            return spc_ch + value

        elif value_type == "int":
            return spc_ch * 2 + str(value)

        elif value_type == "bool":
            value = "T" if value else "F"

        elif value_type == "NoneType":
            value = "N"

        elif value_type == "float":

            value = str(value)
            if value == 'nan':
                value = 'NAN'
            elif value == 'inf':
                value = 'INF'
            elif value == '-inf':
                value = 'NIF'
            elif spc_ch == "_": # for url-mode because float contain dot-inside
                return '-' + HELML.base64Uencode(value)

        return spc_ch * 2 + str(value)


    @staticmethod
    def valueDecoder(encodedValue: str, spc_ch: str = ' ') -> Union[str, int, float, bool, None]:
        """
        Decodes a HELML formatted string into its original value.

        Args:
            encoded_value (str): The HELML encoded value as a string.
            spc_ch (str, optional): The space character used for decoding. Defaults to ' '.

        Returns:
            Union[str, int, float, bool, None]: The decoded value, which can be of type str, int, float, bool, or None.
        """

        # calc stpos by first-2-chars of encodedValue
        #stpos = 2 if encodedValue.startswith(spc_ch*2) else 1 if encodedValue.startswith(spc_ch) else 0
        stpos = 0
        while encodedValue.startswith(spc_ch, stpos):
            stpos += 1

        # raw
        if stpos == 1:
            return encodedValue[1:]

        # special 0
        if stpos == 0:
            fc = encodedValue[0]

            if fc == '-':
                return HELML.base64Udecode(encodedValue[1:])
            elif fc == "'":
                return encodedValue[1:-1]
            elif fc == '"':
                return encodedValue[1:-1].encode('utf-8').decode('unicode_escape')
            elif fc == '%':
                return HELML.hexDecode(encodedValue[1:])

        sliced_value = encodedValue[stpos:]

        # stpos != 1
        if stpos > 1:
            # content-dependent decoding
            if sliced_value in HELML.SPEC_TYPE_VALUES:
                return HELML.SPEC_TYPE_VALUES[sliced_value]

            elif HELML.is_numeric(sliced_value):
                # it's probably a numeric value
                if '.' in sliced_value:
                    # if there's a decimal point, it's a floating point number
                    return float(sliced_value)
                else:
                    # if there's no decimal point, it's an integer
                    return int(sliced_value, 10)

        # if there are no spaces or quotes or "-" at the beginning
        if HELML.CUSTOM_FORMAT_DECODER is not None:
            # use custom-decoder if defined
            return HELML.CUSTOM_FORMAT_DECODER(encodedValue, spc_ch)

        # fallback: return "as is"
        return encodedValue

    @staticmethod
    def base64Uencode(string: str) -> str:
         enc = base64.b64encode(string.encode())
         return enc.decode().rstrip("=").translate(str.maketrans('+/', '-_'))

    @staticmethod
    def base64Udecode(s: str) -> str:
        l = len(s)
        if not l:
            return ""
        l %= 4
        if l:
            s += "=" * (4 - l)
        try:
            decoded = base64.b64decode(s.translate(str.maketrans('-_', '+/')).encode()).decode()
            return decoded
        except Exception:
            return None
        
    @staticmethod
    def hexDecode(encoded: str) -> Union[str, None]:
        hexc1 = '0123456789abcdefABCDEF'
        hexc2 = hexc1 + ' '
        decoded = ''
        i = 0
        while i < len(encoded):
            fc = encoded[i]
            sc = encoded[i+1] if i+1 < len(encoded) else None
            if fc in hexc1 and sc in hexc2:
                decoded += chr(int(fc + sc, 16))
                i += 2
            else:
                return None
        return decoded

    @staticmethod
    def is_numeric(value: str) -> bool:
        try:
            float(value)
            return True
        except ValueError:
            return False

    @staticmethod
    def num_keys_cnt(arr: Union[dict, list, tuple, set]) -> int:
        el_count = len(arr)
        if not el_count:
            return 0
        if not isinstance(arr, dict):
            return el_count
        
        # check for non-empty dict only, get keys list
        keys_list = arr.keys()
        exp_keys_range = range(0, el_count)
        for i in keys_list:
            if not isinstance(i, (int, str)):
                return 0
            try:
                if int(i) not in exp_keys_range:
                    return 0
            except ValueError:
                return 0

        return el_count